src/main/java/org/checkerframework/specimin/Slicer.java

```diff
@@ -0,0 +1,212 @@
+      }
+    }
+
+    // Step 3: add all used compilation units
+    for (CompilationUnit used : usedCompilationUnits) {
+      toSlice.put(
+          qualifiedNameToFilePath(used.getPrimaryType().get().getFullyQualifiedName().get())
+              .toString(),
+          used);
+    }
+
+    // Step 4: remove all nodes except the ones marked for preservation OR package/import
+    // declarations
+    for (CompilationUnit cu : toSlice.values()) {
+      sliceCompilationUnit(cu);
+    }
+  }
+
+  public Set<UnsolvedSymbolAlternates<?>> getGeneratedSymbolSlice() {
+    return generatedSymbolSlice;
+  }
+
+  public void addToWorklist(Node node) {
+    worklist.add(node);
+  }
+
+  private void handleElement(Node node, Set<CompilationUnit> usedCompilationUnits) {
+    if (slice.contains(node)) {
+      return;
+    }
+
+    if (node instanceof Resolvable<?>) {
+      Resolvable<?> asResolvable = (Resolvable<?>) node;
+
+      try {
+        Object resolved = asResolvable.resolve();
+
+        if (resolved == null) {
+          throw new RuntimeException("Unexpected null value in resolve() call");
+        }
+
+        List<Node> toAddToWorklist = typeRuleDependencyMap.getRelevantElements(resolved);
+        addToBeginningOfWorklist(toAddToWorklist);
+
+        // Since resolved declarations may reference another file, we need to add that compilation
+        // unit to the output
+        usedCompilationUnits.addAll(
+            toAddToWorklist.stream().map(n -> n.findCompilationUnit().get()).toList());
+      } catch (UnsolvedSymbolException ex) {
+        // Generate a synthetic type
+        generatedSymbolSlice.addAll(unsolvedSymbolGenerator.inferContext(node));
+      }
+    }
+
+    slice.add(node);
+    addToBeginningOfWorklist(typeRuleDependencyMap.getRelevantElements(node));
+
+    if (unsolvedSymbolGenerator.needToPostProcess(node)) {
+      postProcessingWorklist.add(node);
+    }
+  }
+
+  /**
+   * Adds all nodes to the beginning of the worklist. This ensures that synthetic types are
+   * generated before they are used. Use instead of worklist.addAll.
+   *
+   * @param c The list of nodes.
+   */
+  private void addToBeginningOfWorklist(List<Node> c) {
+    for (int i = c.size(); i > 0; i--) {
+      worklist.addFirst(c.get(i - 1));
+    }
+  }
+
+  /**
+   * Starts the slice on a compilation unit. Use this instead of {@link #sliceNode(Node)} because
+   * {@link CompilationUnit} has some quirks that prevents {@link CompilationUnit#getChildNodes()}
+   * from accessing anything but package/import declarations.
+   *
+   * @param cu The compilation unit
+   */
+  private void sliceCompilationUnit(CompilationUnit cu) {
+    List<TypeDeclaration<?>> typesCopy = new ArrayList<>(cu.getTypes());
+    for (TypeDeclaration<?> typeDecl : typesCopy) {
+      sliceNode(typeDecl);
+    }
+  }
+
+  /**
+   * Recursively removes all nodes not in the slice.
+   *
+   * @param node The node to slice
+   */
+  private void sliceNode(Node node) {
```

Comment:
I'd prefer that we reserve the term "slice" for the overall process, and use a different term for the internal step of removing nodes that this method implements.